File: app/javascript/src/agenda_store/recurrence.js
// Recurrence expander — JS port of AgendaSchedule#matches? + the
// occurrence_date math on the Rails side. Pure functions, no DOM, no
// timezone state. Lets the FE expand a serialize_for_client'd schedule
// into phantom occurrences for any date without a server round-trip.
//
// Any change to the recurrence rule on the Ruby side MUST be ported here
// and verified by spec/javascript/recurrence_parity_spec.rb — that spec
// generates fixtures from Ruby, feeds them to this module via Node, and
// asserts identical output.
//
// Written as CommonJS (`module.exports = …`) so this exact file is
// require()-able from Node for the parity spec AND bundleable by esbuild
// for the browser (esbuild handles both CommonJS and ESM).

const WEEKDAY_KEYS = ["sun", "mon", "tue", "wed", "thu", "fri", "sat"];

// Parse "YYYY-MM-DD" into a date-shaped record. All math is done against
// these records (not Date objects with implicit timezones) so the same
// input always produces the same answer regardless of the host timezone.
function parseISO(iso) {
  const [y, m, d] = String(iso).split("-").map(Number);
  // UTC noon gives a stable weekday calculation across every TZ Node
  // could be launched in — DST shifts can't move the wall date away from
  // its calendar position when we're already inside the day.
  const dt = new Date(Date.UTC(y, m - 1, d, 12, 0, 0));
  return { y, m, d, wday: dt.getUTCDay(), iso: String(iso) };
}

function pad2(n) { return String(n).padStart(2, "0"); }

function formatISO(y, m, d) {
  return `${y}-${pad2(m)}-${pad2(d)}`;
}

function addDays(iso, n) {
  const { y, m, d } = parseISO(iso);
  const dt = new Date(Date.UTC(y, m - 1, d + n, 12, 0, 0));
  return formatISO(dt.getUTCFullYear(), dt.getUTCMonth() + 1, dt.getUTCDate());
}

function dayDiff(fromISO, toISO) {
  const a = parseISO(fromISO);
  const b = parseISO(toISO);
  return Math.round((Date.UTC(b.y, b.m - 1, b.d) - Date.UTC(a.y, a.m - 1, a.d)) / 86400000);
}

function monthsBetween(fromISO, toISO) {
  const a = parseISO(fromISO);
  const b = parseISO(toISO);
  return (b.y - a.y) * 12 + (b.m - a.m);
}

function endOfMonthDay(iso) {
  const { y, m } = parseISO(iso);
  // Day 0 of NEXT month is the last day of THIS month.
  const last = new Date(Date.UTC(y, m, 0, 12, 0, 0));
  return last.getUTCDate();
}

// Public API ---------------------------------------------------------------

function matches(schedule, dateISO) {
  if (!schedule || !schedule.starts_on) return false;
  if (dateISO < schedule.starts_on) return false;
  if (schedule.until_on && dateISO > schedule.until_on) return false;
  const excluded = schedule.excluded_dates || [];
  if (excluded.indexOf(dateISO) !== -1) return false;

  switch (schedule.freq || "daily") {
    case "daily":    return true;
    case "weekdays": return matchesWeekdays(dateISO);
    case "weekly":   return matchesWeekly(schedule, dateISO);
    case "monthly":  return matchesMonthly(schedule, dateISO);
    case "yearly":   return matchesYearly(schedule, dateISO);
    case "custom":   return matchesCustom(schedule, dateISO);
    default:         return false;
  }
}

// Materialize every matching date in [fromISO..toISO] (inclusive). Pure
// list; the AgendaStore wires this up with buildPhantom + a suppression
// map for materialized overrides.
function expand(schedule, fromISO, toISO) {
  if (!schedule || !schedule.starts_on) return [];
  // Trim against the schedule's own active window so we never iterate
  // years we know can't match.
  const lower = fromISO < schedule.starts_on ? schedule.starts_on : fromISO;
  const upper = schedule.until_on && toISO > schedule.until_on ? schedule.until_on : toISO;
  if (upper < lower) return [];

  const out = [];
  let cur = lower;
  // Hard cap on iteration: 50 years of days. Way past any sensible
  // window, but a guard against a malformed rule running away.
  const MAX = 50 * 366;
  let count = 0;
  while (cur <= upper && count < MAX) {
    if (matches(schedule, cur)) out.push(cur);
    cur = addDays(cur, 1);
    count += 1;
  }
  return out;
}

// Build an AgendaItem-shaped phantom for `dateISO`. Caller must pass a
// `localEpoch(dateISO, "HH:MM") → seconds_since_epoch` function that
// honors the user's timezone — see ./timezone.js. The optional `agenda`
// lookup lets us populate the agenda-name/color/source presentation
// attrs so the renderer fills the row exactly like a materialized one.
function buildPhantom(schedule, dateISO, { localEpoch, agenda }) {
  const startEpoch = localEpoch(dateISO, schedule.start_time);
  const endEpoch = (schedule.kind === "event" && schedule.duration_minutes)
    ? startEpoch + (Number(schedule.duration_minutes) * 60)
    : null;
  const id = `p-${schedule.id}-${dateISO}`;
  const color = schedule.color || (agenda && agenda.color) || "";
  // Travel metadata is inherited from the schedule's nested `metadata.travel`
  // hash — populated by AgendaTravelChain::Service for the parent schedule.
  // Without this, every phantom rendered a 0-minute band even when the
  // parent schedule's resolver had landed a real travel time (e.g. the
  // recurring TMS event showed the band on today's materialized row but
  // not on Friday's phantom).
  const meta = (schedule && schedule.metadata) || {};
  const travel = (meta && meta.travel) || {};
  // Project the leg array (server sends [{to, drive_seconds, dwell_seconds, ...}])
  // down to the minimal shape the renderer needs. Returning null collapses
  // to the empty-string presentation attr — recurrence schedules without a
  // multi-stop chain stay byte-identical to before.
  const legPayload = (legs) => Array.isArray(legs) && legs.length > 1
    ? legs.map((leg) => ({
        to:             leg.to || "",
        drive_seconds:  Number(leg.drive_seconds) || 0,
        dwell_seconds:  Number(leg.dwell_seconds) || 0,
      }))
    : null;
  // Mirrors `AgendaItem#presentation_attrs` — the renderer reads from
  // this hash; without it phantoms render as blank rows (just the
  // template skeleton: checkbox + edit pencil, no name/time/anything).
  const presentation_attrs = {
    "item-id":              id,
    "item-url":             `/agenda_items/${id}`,
    "phantom":              true,
    "recurring":            true,
    "agenda-schedule-id":   schedule.id,
    "detached":             false,
    "kind":                 schedule.kind,
    "color":                color,
    "agenda-id":            schedule.agenda_id,
    "agenda-name":          agenda ? agenda.name : "",
    "agenda-color":         agenda ? agenda.color : "",
    "agenda-source":        agenda ? agenda.source : "",
    "all-day":              !!schedule.all_day,
    // `end-date` is the INCLUSIVE last-day-midnight epoch (mirrors
    // `AgendaItem#presentation_attrs` and `optimistic_item.js`). For an
    // all-day phantom, `endEpoch` is the exclusive next-day-midnight
    // (start + duration_minutes*60, where Google-synced all-day events
    // always set duration to a multiple of 1440). Walk back one day so
    // the cal_week / cal_month banner layout doesn't render a single-day
    // all-day event across two columns. Was: `endEpoch || startEpoch` —
    // missed the walk-back and made every Google-synced recurring all-day
    // event bleed into the next day's column.
    "end-date":             (!!schedule.all_day && endEpoch) ? (endEpoch - 86400) : (endEpoch || startEpoch),
    "start-at":             startEpoch,
    "end-at":               endEpoch,
    "name":                 schedule.name || "",
    "notes":                schedule.notes || "",
    "location":             schedule.location || "",
    "resolved-address":     travel.location_address || "",
    "arrive-early-minutes": Number(schedule.arrive_early_minutes) || 0,
    "travel-minutes":       Number(travel.travel_minutes) || 0,
    "travel-from-kind":     travel.travel_from_kind || "",
    "travel-from":          travel.travel_from || "",
    "chain-predecessor-id": travel.chain_predecessor_id || "",
    "chain-successor-id":   travel.chain_successor_id || "",
    "chain-prev-end-epoch": travel.chain_prev_end_at || "",
    "leave-at-epoch":       travel.leave_at || "",
    "post-travel-to":       travel.post_travel_to || "",
    "post-travel-minutes":  Number(travel.post_travel_minutes) || 0,
    "post-arrive-at-epoch": travel.post_arrive_at || "",
    "before-legs":          travel.before_legs ? JSON.stringify(legPayload(travel.before_legs)) : "",
    "after-legs":           travel.after_legs ? JSON.stringify(legPayload(travel.after_legs)) : "",
    "trigger-expression":   schedule.trigger_expression || "",
    "schedule":             JSON.stringify(schedule),
    "attendees":            "[]",
    "organizer":            "null",
    "self-response":        "",
  };
  return {
    id:                   id,
    agenda_id:            schedule.agenda_id,
    agenda_name:          agenda ? agenda.name : "",
    agenda_color:         agenda ? agenda.color : "",
    agenda_schedule_id:   schedule.id,
    kind:                 schedule.kind,
    name:                 schedule.name,
    notes:                schedule.notes,
    location:             schedule.location,
    color:                color,
    all_day:              !!schedule.all_day,
    arrive_early_minutes: Number(schedule.arrive_early_minutes) || 0,
    trigger_expression:   schedule.trigger_expression,
    metadata:             schedule.metadata || {},
    start_at:             startEpoch,
    end_at:               endEpoch,
    phantom:              true,
    recurring:            true,
    detached:             false,
    status:               "confirmed",
    completed_at:         null,
    attendees:            [],
    organizer:            null,
    self_response:        "",
    editable:             agenda ? agenda.editable !== false : true,
    schedule:             schedule,
    occurrence_date:      dateISO,
    presentation_attrs:   presentation_attrs,
  };
}

// Frequency helpers --------------------------------------------------------

function matchesWeekdays(dateISO) {
  const { wday } = parseISO(dateISO);
  return wday >= 1 && wday <= 5;
}

function matchesWeekly(schedule, dateISO) {
  const { wday } = parseISO(dateISO);
  return weekdayIndices(schedule).indexOf(wday) !== -1;
}

function weekdayIndices(schedule) {
  const fromRule = (schedule.by_day || [])
    .map((k) => WEEKDAY_KEYS.indexOf(String(k).toLowerCase()))
    .filter((i) => i >= 0);
  if (fromRule.length === 0) return [parseISO(schedule.starts_on).wday];
  return fromRule;
}

function matchesMonthly(schedule, dateISO) {
  // "Nth weekday of month" (e.g. third Tuesday) takes precedence when
  // both by_set_pos and by_day are set — matches Ruby#matches_month_day?.
  if (schedule.by_set_pos && (schedule.by_day || []).length > 0) {
    return matchesNthWeekdayOfMonth(schedule, dateISO);
  }
  const { d } = parseISO(dateISO);
  const days = monthDays(schedule);
  if (days.indexOf(d) !== -1) return true;
  if (days.indexOf(-1) !== -1 && d === endOfMonthDay(dateISO)) return true;
  return false;
}

function monthDays(schedule) {
  const arr = (schedule.by_month_day || []).map((n) => Number(n)).filter((n) => !Number.isNaN(n));
  if (arr.length === 0) return [parseISO(schedule.starts_on).d];
  return arr;
}

function matchesYearly(schedule, dateISO) {
  const start = parseISO(schedule.starts_on);
  const here = parseISO(dateISO);
  return start.m === here.m && start.d === here.d;
}

function matchesCustom(schedule, dateISO) {
  const interval = Math.max(1, Number(schedule.interval) || 1);
  const unit = String(schedule.unit || "day").toLowerCase();
  switch (unit) {
    case "day": {
      return dayDiff(schedule.starts_on, dateISO) % interval === 0;
    }
    case "week": {
      const diff = dayDiff(schedule.starts_on, dateISO);
      if (Math.floor(diff / 7) % interval !== 0) return false;
      return parseISO(schedule.starts_on).wday === parseISO(dateISO).wday;
    }
    case "month": {
      const months = monthsBetween(schedule.starts_on, dateISO);
      if (months % interval !== 0) return false;
      if (schedule.by_set_pos && (schedule.by_day || []).length > 0) {
        return matchesNthWeekdayOfMonth(schedule, dateISO);
      }
      if ((schedule.by_month_day || []).length > 0) {
        return matchesMonthly(schedule, dateISO);
      }
      return parseISO(schedule.starts_on).d === parseISO(dateISO).d;
    }
    default: return false;
  }
}

function matchesNthWeekdayOfMonth(schedule, dateISO) {
  const setPos = Number(schedule.by_set_pos);
  const targetKey = (schedule.by_day || [])[0];
  if (!targetKey) return false;
  const targetWday = WEEKDAY_KEYS.indexOf(String(targetKey).toLowerCase());
  if (targetWday < 0) return false;
  const { y, m, d, wday } = parseISO(dateISO);
  if (wday !== targetWday) return false;

  if (setPos === -1) {
    // last weekday of month: adding 7 days moves into the next month
    const probe = new Date(Date.UTC(y, m - 1, d + 7, 12, 0, 0));
    return (probe.getUTCMonth() + 1) !== m;
  }
  const weekOfMonth = Math.floor((d - 1) / 7) + 1;
  return weekOfMonth === setPos;
}

const Recurrence = {
  matches,
  expand,
  buildPhantom,
  parseISO,
  addDays,
  dayDiff,
};

// CommonJS export so Node can `require` this directly in the parity
// spec without a build step. esbuild also accepts this form and inlines
// it cleanly at bundle time for the browser.
if (typeof module !== "undefined" && module.exports) module.exports = Recurrence;
if (typeof window !== "undefined") window.AgendaRecurrence = Recurrence;

File: app/service/recurrence.rb
# "Every weekday", "the second Tuesday of the month", "every other Thursday".
#
# A pure value object over a recurrence rule and the window it runs in. No
# database, no timezone state, no knowledge of what's recurring - it answers
# whether a DATE is in the pattern, and walks forward to the next one that is.
#
# Lifted out of AgendaSchedule, which owned this and still holds every other
# concern a calendar series has (materializing rows, phantoms, Google sync).
# BuddyReminder needed the same rules and had a fraction of them - four
# frequencies with no interval, no nth-weekday, no end date - so "remind me
# every second Tuesday" was a thing the calendar could express and a reminder
# could not, for no reason other than which file the code happened to live in.
#
# The rule is the shape AgendaSchedule already stored, so nothing about the
# calendar's data had to change:
#
#   { freq: :weekly,  by_day: [:mon, :wed] }
#   { freq: :monthly, by_month_day: [1, 15] }
#   { freq: :monthly, by_set_pos: 2, by_day: [:tue] }   # second Tuesday
#   { freq: :monthly, by_set_pos: -1, by_day: [:fri] }  # last Friday
#   { freq: :custom,  interval: 2, unit: :week }        # every other <starts_on's weekday>
#
# There is a JS port of this at app/javascript/src/agenda_store/recurrence.js
# and spec/javascript/recurrence_parity_spec.rb runs both over the same rules
# and diffs the results. A new branch here MUST be ported there in the same
# change, or the calendar renders different days depending on whether the page
# hit the server.
class Recurrence
  FREQUENCIES  = %i[daily weekdays weekly monthly yearly custom].freeze
  WEEKDAY_KEYS = %i[sun mon tue wed thu fri sat].freeze
  CUSTOM_UNITS = %i[day week month].freeze

  # How far `next_on_or_after` will walk before giving up. A yearly rule needs
  # 366 in the worst case and a custom monthly interval of 12 needs about the
  # same; five years is slack on both and bounds a rule that matches nothing at
  # all (a `weekly` with an empty `by_day` that somehow got past validation)
  # rather than hanging the request that asked.
  SCAN_LIMIT_DAYS = 5 * 366

  # Resolving "stop after N times" into a concrete end date. 50 years is far
  # past any real schedule and keeps a bad rule from looping forever.
  OCCURRENCE_SCAN_CAP = 50.years

  attr_reader :starts_on, :until_on

  def initialize(rule, starts_on:, until_on: nil)
    @rule      = (rule || {}).with_indifferent_access
    @starts_on = starts_on&.to_date
    @until_on  = until_on&.to_date
  end

  def freq
    (@rule[:freq].to_s.presence || :daily).to_sym
  end

  def valid_freq?
    FREQUENCIES.include?(freq)
  end

  def interval = [@rule[:interval].to_i, 1].max
  def set_pos  = @rule[:by_set_pos].presence&.to_i

  def unit
    given = (@rule[:unit].to_s.presence || :day).to_sym
    CUSTOM_UNITS.include?(given) ? given : :day
  end

  def by_day
    Array(@rule[:by_day]).map { |d| d.to_s.downcase }
  end

  def excluded_dates
    Array(@rule[:excluded_dates]).filter_map { |d| Date.parse(d.to_s) rescue nil }.to_set
  end

  # Is `date` an occurrence? Answers the whole question, end date included.
  def matches?(date)
    date = date.to_date
    return false if until_on.present? && date > until_on

    matches_rule?(date)
  end

  # The pattern alone, ignoring the end date. Split out because resolving
  # "stop after N" walks the rule to DERIVE the end date, and consulting a
  # bound that doesn't exist yet would cut the walk short.
  def matches_rule?(date)
    date = date.to_date
    return false if starts_on.nil? || date < starts_on
    return false if excluded_dates.include?(date)

    case freq
    when :daily    then true
    when :weekdays then (1..5).cover?(date.wday)
    when :weekly   then weekday_indices.include?(date.wday)
    when :monthly  then matches_month_day?(date)
    when :yearly   then date.month == starts_on.month && date.day == starts_on.day
    when :custom   then matches_custom?(date)
    else                false
    end
  end

  # First occurrence on or after `date`, or nil if the rule runs out first.
  def next_on_or_after(date)
    cursor = [date.to_date, starts_on].compact.max
    return nil if cursor.nil?

    SCAN_LIMIT_DAYS.times do
      return nil if until_on.present? && cursor > until_on
      return cursor if matches?(cursor)

      cursor += 1
    end
    nil
  end

  # Every occurrence in [from..to]. Bounded by the rule's own window so a
  # request for the next decade doesn't iterate years that can't match.
  def between(from, to)
    lower = [from.to_date, starts_on].compact.max
    upper = until_on.present? ? [to.to_date, until_on].min : to.to_date
    return [] if lower.nil? || upper < lower

    (lower..upper).select { |date| matches?(date) }
  end

  # The date the Nth occurrence lands on, for "stop after 10 times". Returns
  # nil when the rule runs dry first.
  def date_of_occurrence(count)
    remaining = count.to_i
    return nil if remaining <= 0 || starts_on.nil?

    date = starts_on
    cap  = starts_on + OCCURRENCE_SCAN_CAP
    last = nil
    while date <= cap && remaining.positive?
      if matches_rule?(date)
        last = date
        remaining -= 1
      end
      date += 1
    end
    last
  end

  private

  def weekday_indices
    by_day.filter_map { |d| WEEKDAY_KEYS.index(d.to_sym) }.presence || [starts_on.wday]
  end

  def month_days
    Array(@rule[:by_month_day]).map(&:to_i).presence || [starts_on.day]
  end

  def matches_month_day?(date)
    # Monthly + Nth weekday - "third Tuesday of every month". When the rule
    # carries both `by_set_pos` and `by_day`, `by_month_day` is ignored.
    return matches_nth_weekday_of_month?(date) if set_pos.present? && by_day.any?

    month_days.include?(date.day) || (month_days.include?(-1) && date.day == date.end_of_month.day)
  end

  def matches_custom?(date)
    case unit
    when :day   then ((date - starts_on).to_i % interval).zero?
    when :week  then (((date - starts_on).to_i / 7) % interval).zero? && date.wday == starts_on.wday
    when :month then matches_custom_month?(date)
    end
  end

  def matches_custom_month?(date)
    return false unless (months_between(starts_on, date) % interval).zero?
    return matches_nth_weekday_of_month?(date) if set_pos.present? && by_day.any?
    return matches_month_day?(date) if Array(@rule[:by_month_day]).any?

    date.day == starts_on.day
  end

  # "Second Thursday", "last Friday". `set_pos` is 1..4 or -1 for last.
  def matches_nth_weekday_of_month?(date)
    target = WEEKDAY_KEYS.index(by_day.first.to_s.to_sym)
    return false if target.nil? || date.wday != target
    return (date + 7).month != date.month if set_pos == -1

    (((date.day - 1) / 7) + 1) == set_pos
  end

  def months_between(from, to)
    ((to.year - from.year) * 12) + (to.month - from.month)
  end
end
